### DISCUSSION TITLE
@fastify/deepmerge

### DISCUSSION THREAD

**Discussion by Uzlopak** [RebeccaStevens/deepmerge-ts#160] [General]
Hi @RebeccaStevens 

I also reimplemented deepmerge at fastify/deepmerge. It is much faster than the original deepmerge. It lacks the support of merging Sets and Maps, but maybe you have some thoughts about the implementation or the typings. Maybe you also want to improve the performance of your implementation and modify your code depending on it.

**Comment by RebeccaStevens** [comment]
Hey @Uzlopak, I've had a look at that library.

With regard to typings, I had to write them for my library completely independently of the code itself as TypeScript wasn't able to infer everything based purely off simple function signatures. My code base ended up being more types than actual implementation code. You'll probably need to do a similar thing if you want to get your types right. (Though you can always just take a more simple approach like this: [v4_0.d.ts](https://github.com/RebeccaStevens/deepmerge-ts/blob/main/types-legacy/v4_0.d.ts#L59-L70))

However getting the types right for the case when users pass in custom configuration data, specifically custom functions, was a real challenge. I managed to eventually get something working with the help of HKTs of the same implementation as [fp-ts's HKT](https://gcanti.github.io/fp-ts/modules/HKT.ts.html). If you're going to add support for this sort of thing to you're library then I wish you luck.

With regards to the implementation itself, you code seems fine but I would recommend adding an option to not clone data and instead just use it as is. There are many use cases where users may want the leaves to the trees to be left alone. This will also end up speeding up your implementation as there is less work for the library to do.

In my library, I don't even give an option for users to clone data; if they want that then need to integrate a deep clone library of their choice in with this library (using `deepmergeCustom`). There were several reason I decided to go with this approach; one being that for any arbitrary object, I don't know how it should be cloned. If it is a Record (aka a plain object) then it's pretty straight forward, but with other objects such as a `Date`, there's no point cloning them and with a custom `Foo` object, I have no idea how to clone it. This clone issue is also currently plaguing the original deepmerge library.

I'm not sure if there is anything I can really take out of your library an integrate here as the two take different approaches. I believe the main reason why yours is much faster is because it is making assumptions about the inputs (such as it's not going to be some `Foo` type) and thus doesn't need to do as many sanity checks and branching.

**Comment by RebeccaStevens** [comment]
@Uzlopak Just a heads-up: v8.0.0 of deepmerge-ts has shipped and brings many improvements, especially in performance.

- **Faster type checking** — the type-level work dropped from 6.5M instantiations / 10.5s down to 83k / 0.94s on the wide benchmark (roughly -90% / -98.7%).
- **Faster runtime** — a dedicated fast path for merging two records alone made two-object merges ~40% faster vs v7.1.6, and memory use dropped dramatically on large multi-object merges.

@fastify/deepmerge now trails deepmerge-ts by a comfortable ~2,000-4,000x according to our benchmarking. Let me know if our benchmark isn't a fair representation and I can fix it.
